### PR TITLE
support cluster events

### PR DESCRIPTION
**Pull request by dongluochen**
This PR is ready for review.

===
This is a work item in progress. Test needs to be added. It also needs a change in Swarmkit Watch API https://github.com/docker/swarmkit/pull/2099. Initial design discussion is at https://github.com/docker/swarmkit/issues/491. I'd like to get design feedback. 

Signed-off-by: Dong Chen <dongluo.chen@docker.com>

**- What I did**
Add swarm events to docker event stream. 

**- How I did it**
Use Swarm Store Watch API to get change notification from Raft store. Translate that into Docker event format and push it into Docker event structure. 

**- How to verify it**
On a manager node, `docker events` includes node's local events and cluster events. Cluster events have a global scope while node's local events are local. Existing event filters apply to cluster events. A new `scope` filter is added. 

```
#run this on a manager node
$ docker events -f scope=global

# a node joins a cluster
2017-04-06T18:03:46.551104594Z node create s0ugk1wi0vgxnspxx0ptmon47 (name=)
2017-04-06T18:03:46.553642227Z node update s0ugk1wi0vgxnspxx0ptmon47 (name=)
2017-04-06T18:03:46.556070184Z node update s0ugk1wi0vgxnspxx0ptmon47 (name=)
2017-04-06T18:03:46.562025609Z node update s0ugk1wi0vgxnspxx0ptmon47 (name=)
2017-04-06T18:03:46.689608127Z node update s0ugk1wi0vgxnspxx0ptmon47 (name=ip-172-19-71-145, state.new=ready, state.old=unknown)
# a node goes down
2017-04-06T18:04:32.705082118Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51, state.new=down, state.old=ready)
# a node is back up
2017-04-06T18:05:06.288643169Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51, state.new=ready, state.old=down)
# promote a node
2017-04-06T18:05:29.965972204Z node update 0urgktyobae3etk5e4n4331es (desiredrole.new=manager, desiredrole.old=worker, name=ip-172-19-147-51)
2017-04-06T18:05:29.972791974Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:05:29.992599632Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:05:30.007889135Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:05:30.084421029Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:05:30.088454689Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
# demote a node
2017-04-06T18:06:15.004097376Z node update 0urgktyobae3etk5e4n4331es (desiredrole.new=worker, desiredrole.old=manager, name=ip-172-19-147-51)
2017-04-06T18:06:20.015726988Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:06:20.035951960Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:06:20.048380156Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:06:20.068198599Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
2017-04-06T18:06:20.072033661Z node update 0urgktyobae3etk5e4n4331es (name=ip-172-19-147-51)
# change a node's availability to pause
2017-04-06T18:07:01.620569322Z node update 0urgktyobae3etk5e4n4331es (availability.new=pause, availability.old=active, name=ip-172-19-147-51)
# change a node's availability to active
2017-04-06T18:07:35.825859057Z node update 0urgktyobae3etk5e4n4331es (availability.new=active, availability.old=pause, name=ip-172-19-147-51)

# create a service
2017-04-06T18:08:43.303340796Z service create 9vvofszhb6iv4k3tmphras96u (name=nginx)
2017-04-06T18:08:43.307625739Z service update 9vvofszhb6iv4k3tmphras96u (name=nginx)
# scale a service
2017-04-06T18:09:37.018798236Z service update 9vvofszhb6iv4k3tmphras96u (name=nginx, replicas.new=3, replicas.old=2)
# update image of a service
2017-04-06T18:11:19.122732546Z service update 9vvofszhb6iv4k3tmphras96u (image.new=nginx:1.10.3@sha256:6202beb06ea61f44179e02ca965e8e13b961d12640101fca213efbfd145d7575, image.old=nginx:latest@sha256:e6693c20186f837fc393390135d8a598a96a833917917789d63766cab6c59582, name=nginx)
2017-04-06T18:11:19.126619069Z service update 9vvofszhb6iv4k3tmphras96u (name=nginx, updatestate.new=updating, updatestate.old=nil)
2017-04-06T18:11:41.552581741Z service update 9vvofszhb6iv4k3tmphras96u (name=nginx, updatestate.new=completed, updatestate.old=updating)
# remove a service
2017-04-06T18:12:23.466026101Z service remove 9vvofszhb6iv4k3tmphras96u (name=nginx)

# create a network
2017-04-06T18:12:45.047497468Z network create qew8jjh6riv75r8tj4wf0imfw (name=tnet)
2017-04-06T18:12:45.053957928Z network update qew8jjh6riv75r8tj4wf0imfw (name=tnet)
# create a container associtated with this network
# create a service associated with this network
2017-04-06T18:14:38.161988206Z service create v6md2fr205au7c6wq5zdo6sz1 (name=nginx)
2017-04-06T18:14:38.166951070Z service update v6md2fr205au7c6wq5zdo6sz1 (name=nginx)
# remove the service
2017-04-06T18:15:29.441350106Z service remove v6md2fr205au7c6wq5zdo6sz1 (name=nginx)
# remove a network
2017-04-06T18:15:43.558832104Z network remove qew8jjh6riv75r8tj4wf0imfw (name=tnet)

# create a secret
2017-04-06T18:16:27.127307662Z secret create p3z8b3f2q40fun94yq3vd2g9y (name=mysecret)
# remove a secret
2017-04-06T18:16:55.393649748Z secret remove p3z8b3f2q40fun94yq3vd2g9y (name=mysecret)
```

**- Description for the changelog**
Add cluster events to Docker event stream.

### CODE REVIEW

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Maybe we should create a `func eventTimestamp(swarmapi.Meta, swarmapi.WatchActionKind) time.Time` function that handles extracting the timestamp. Then we can call it in `generateClusterEvent` and pass in the timestamp instead of needing to handle it in all the different switch statements.

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: Would it be useful for `Watch` to support something like `Kind: "*"`?

In general I want people to filter for the specific events of interest, but the Docker events API is a bit of an unusual case.

Comment (dongluochen) [inline daemon/cluster/noderunner.go]: We don't want to get `task` change notifications. They are too many and are implementation details that we don't user rely on that.

**Comment by dongluochen** [comment]
Test failure looks unrelated. It's already tracked by https://github.com/moby/moby/issues/33041. 

```
23:20:09 ----------------------------------------------------------------------
23:20:09 FAIL: check_test.go:355: DockerSwarmSuite.TearDownTest
23:20:09 
23:20:09 check_test.go:360:
23:20:09     d.Stop(c)
23:20:09 daemon/daemon.go:392:
23:20:09     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
23:20:09 ... Error: Error while stopping the daemon d95be62fe1bc6 : exit status 2
```

**Comment by dongluochen** [comment]
Please review.

File: daemon/events.go
Comment (AkihiroSuda) [inline daemon/events.go]: logrus.Warn on default?

File: daemon/events.go
Comment (AkihiroSuda) [inline daemon/events.go]: Can we use make these string literals `const`-ants?

**Comment by AkihiroSuda** [comment]
cc @aluzzardi @aaronlehmann

File: daemon/cluster/cluster.go
Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: Comment doesn't match the field name.

Maybe this can just be `Events`? I like to avoid putting type information in the field names.

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: I can't see any code path where it is possible for `n.cluster` to be nil. I'd suggest removing this - it is a little misleading to suggest that `n.cluster` can be nil, when other code would blow up in this case.

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: `logrus.WithError(err).Error("failed to receive changes from store watch API")`

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: Should this log anything?

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: Select on this channel write and `<-ctx.Done()`.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Is there any risk of a panic or deadlock if `generateClusterEvent` causes `daemon.EventsService.PublishMessage` to be called during the daemon's shutdown process? This is a separate goroutine that might continue receiving events as the daemon shuts down.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Does it make sense to include `updatestate.old` at all if there is no old update state?

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: `notcontainer` is a strange value. Maybe we could use `Sprintf("%T", oldService.Spec.Task.GetRuntime())` to include the type name. On the other hand, I'm not sure this case is important enough to handle at all.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: `notreplicate` is a strange value. Maybe we could use `Sprintf("%T", oldService.Spec.GetMode())` to include the type name. On the other hand, I'm not sure this case is important enough to handle at all.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Empty case

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Empty case

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Don't we need something to check if `oldNode` is nil? The old value is not guaranteed to be available.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Don't we need something to check if `oldService` is nil? The old value is not guaranteed to be available.

File: daemon/events/events.go
Comment (aaronlehmann) [inline daemon/events/events.go]: "and publishes it"

**Review by aaronlehmann — CHANGES_REQUESTED** [review]

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Providing the old object is best-effort. We're not able to do it if you're looking at past events. I don't think the events code is using this right now, but the point is that the API doesn't guarantee it can provide the old object. Even if it did, this code should still check. It's node good for the client side of a client/server system to crash if the server omits expected values.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: To me, not having an "old" field makes it even more obvious that there is no old state.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Don't we need to check if `OldObject` is nil before calling `GetNode`, etc.?

Comment (dongluochen) [inline daemon/events.go]: The `GetNode` function handles it directly. 

```
135  func (m *Object) GetNode() *Node {
    1   if x, ok := m.GetObject().(*Object_Node); ok {
    2     return x.Node
    3   }
    4   return nil
    5 }
```

**Comment by aaronlehmann** [comment]
LGTM

**Review by aaronlehmann — APPROVED** [review]

**Comment by aaronlehmann** [comment]
Please rebase

**Comment by thaJeztah** [comment]
@dongluochen looks like CI failed on a TearDown, could this be related?

```
21:22:11 
21:22:11 ----------------------------------------------------------------------
21:22:11 FAIL: check_test.go:355: DockerSwarmSuite.TearDownTest
21:22:11 
21:22:11 check_test.go:360:
21:22:11     d.Stop(c)
21:22:11 daemon/daemon.go:392:
21:22:11     t.Fatalf("Error while stopping the daemon %s : %v", d.id, err)
21:22:11 ... Error: Error while stopping the daemon d7e64b157df6f : exit status 2
21:22:11 
21:22:11 
21:22:11 ----------------------------------------------------------------------
21:22:11 PANIC: docker_api_swarm_service_test.go:27: DockerSwarmSuite.TestAPIServiceUpdatePort
21:22:11 
21:22:11 [d7e64b157df6f] waiting for daemon to start
21:22:11 [d7e64b157df6f] daemon started
21:22:11 
21:22:11 [d7e64b157df6f] exiting daemon
21:22:11 ... Panic: Fixture has panicked (see related PANIC)
21:22:11 
21:22:11 ----------------------------------------------------------------------
```

**Comment by aaronlehmann** [comment]
I think it is related.

**Comment by dongluochen** [comment]
Looking at it. I saw similar issue at https://github.com/moby/moby/issues/33041 so I didn't pay attention. I should have looked into it.

File: daemon/events.go
Comment (aaronlehmann) [inline daemon/events.go]: Remove this before merging, or make it a debug log.

Comment (dongluochen) [inline daemon/events.go]: I'll change it to debug.

File: daemon/events.go
Comment (aluzzardi) [inline daemon/events.go]: **Maybe** this should be `swarm`. Not sure though.

@dongluochen @aaronlehmann ?

File: daemon/events.go
Comment (aluzzardi) [inline daemon/events.go]: I think we should output `node.Role` (`attributes["role..."]`) rather than `node.Spec.DesiredRole`

File: daemon/events.go
Comment (dongluochen) [inline daemon/events.go]: I'll update it to `swarm`.

File: daemon/events.go
Comment (dongluochen) [inline daemon/events.go]: Changed to `role.new`.

Comment (aaronlehmann) [inline daemon/events.go]: The code is still using `node.Spec.DesiredRole`. I believe @aluzzardi wanted this to look at `node.Role` instead.

Comment (dongluochen) [inline daemon/events.go]: Thanks @aaronlehmann. Updated.

**Review by aaronlehmann — APPROVED** [review]

**Comment by dongluochen** [comment]
Ping @AkihiroSuda @vdemeester @tonistiigi. Please take a look if you are interested.

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: There shouldn't be a need to close this. I think there is a possibility that cluster can still attempt to write to this chan after `Cleanup` has been called(although very unlikely as `shutdownDaemon` is not fast).

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: No need to change but wondering why you chose to pass in a channel and handle this in dockerd instead of defining a `LogEvent` interface in cluster, that would be called to process events.

Comment (dongluochen) [inline cmd/dockerd/daemon.go]: It was decided to add a `watch` API to expose cluster change. But do not add a separate `LogEvent` interface for simplicity. 
https://github.com/docker/swarmkit/issues/491#issuecomment-282173645

File: cmd/dockerd/daemon.go
Comment (dongluochen) [inline cmd/dockerd/daemon.go]: Thanks. Let me remove it.

Comment (aaronlehmann) [inline cmd/dockerd/daemon.go]: If it is not closed, the `ProcessClusterNotifications` goroutine will never exit.

Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: @aaronlehmann It should take a context then. Or the `cluster.Cleanup()` should close it after it knows there are no writes coming.

File: cmd/dockerd/daemon.go
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: I meant interface in docker's `cluster` pkg like `cluster.Backend` `cluster.NetworkSubnetsProvider` that would be implemented by `daemon.Daemon`

File: cmd/dockerd/daemon.go
Comment (dongluochen) [inline cmd/dockerd/daemon.go]: Add a context to `ProcessClusterNotifications`.

File: cmd/dockerd/daemon.go
Comment (dongluochen) [inline cmd/dockerd/daemon.go]: I'll look at it in a separate change.

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: Does this need to also select on the `Done` method of the context passed to `ProcessClusterNotifications`, so that this won't block forever if `ProcessClusterNotifications` exits due to context cancellation?

Comment (dongluochen) [inline daemon/cluster/noderunner.go]: I don't think it necessary to do that. `c.Cleanup()` at https://github.com/moby/moby/blob/master/cmd/dockerd/daemon.go#L278 would trigger return on this goroutine. The call path is `Cluster.Cleanup() -> nodeRunner.Stop() -> set nodeRunner.closed -> nodeRunner.handleNodeExit() -> cancel nodeRunner.handleControlSocketChange context -> store watch API got cancelled`.

File: daemon/events.go
Comment (vdemeester) [inline daemon/events.go]: We don't display `labels` changes, by design ?

**Review by vdemeester — APPROVED** [review]
LGTM 🦁 😍 
(one small question, but can't wait to have that !!)

File: daemon/cluster/noderunner.go
Comment (aaronlehmann) [inline daemon/cluster/noderunner.go]: > `store watch API got cancelled`

This will never happen if this function is blocked on the channel write instead of calling `Recv`. I don't see what prevents `ProcessClusterNotifications` from exiting before this function.

As an aside, I think it's bad for `ProcessClusterNotifications` to exit on either a channel close or a context cancellation. There should be only one way to shut it down.

Comment (dongluochen) [inline daemon/cluster/noderunner.go]: Let's use the context from `nodeRunner.handleControlSocketChange` to make sure it exits.

File: daemon/events.go
Comment (dongluochen) [inline daemon/events.go]: The `attributes` are extra information for an event. For example, a node update event `2017-04-06T18:07:01.620569322Z node update 0urgktyobae3etk5e4n4331es (availability.new=pause, availability.old=active, name=ip-172-19-147-51)` contains a fix part `timestamp node update nodeID` and the attributes part `availability.new, availability.old, name`. We can add `labels` change to the attributes if they are useful. 

Attributes are added to events in an ad-hoc way. Since there are a lot of information in swarm objects, revealing all changes may reduce readability. For example, a node goes thru a list of changes when joining a cluster, which is internal procedure and users shouldn't put much effort to inspect them. 

I'm not very sure how this will be used. We are starting with basic events https://github.com/docker/swarmkit/issues/491#issuecomment-282446833. I think it'll be extended based on user feedbacks.

Comment (thaJeztah) [inline daemon/events.go]: I've seen many tools using labels to add configuration that's used when listening for events - labels would be a likely candidate to add

**Pull request merged**
